### ISSUE TITLE
HttpContent is encoded two times when posting data

### ISSUE THREAD

**Issue by chingor13** [googleapis/google-http-java-client#648]
The resulting request is correctly encoded once, but we are encoding an extra time when we calculate the content length.

**Comment by sduskis** [comment]
This is a duplicate of #283.

**Comment by chingor13** [comment]
It turns out this is non-trivial to fix. The issue is that for some transport adapters, we have to set the Content-Length header before we start streaming the content. This means that we must count the bytes before we send them. We don't want to cache the encoded content in memory as it defeats the purpose of streaming the content and could cause memory spikes in the application. We don't want to write to a temporary file as there may be file system issues/restrictions (like on GAE) that we cannot necessarily plan for.

**Comment by codyoss** [comment]
@chingor13 do we have more details on this one? Such as what is being affected? Or is there a finite list of Transport Adapters that need the `Content-Length` set?

**Comment by codyoss** [comment]
I will take a stab at this.
